### ISSUE TITLE
File Cache gets EOFError

### ISSUE THREAD

**Issue by csm10495** [pallets/cachelib#21]
I'm seeing a case where sometimes I get:

```
  File "/.../site-packages/cachelib/file.py", line 147, in set
    self._prune()
  File "/.../site-packages/cachelib/file.py", line 96, in _prune
    expires = pickle.load(f)
EOFError: Ran out of input
```

Should this error be in the try/except block?

**Comment by northernSage** [comment]
Hey @csm10495, thanks for taking the time to open the issue. This seems to be happening in [cachelib.FileSystemCache._prune](https://github.com/pallets/cachelib/blob/281fc28b83bd4d2b00e109d9944bcb8937becd92/cachelib/file.py#L96). Now, bear in mind that the pruning logic has seen quite a few changes and a new release should come out soon (hopefully). That being said, `cachelib.FileSystemCache` should not create empty files under normal circumstances, we can see that in [cachelib.FileSystemCache.set](https://github.com/pallets/cachelib/blob/281fc28b83bd4d2b00e109d9944bcb8937becd92/cachelib/file.py#L140) where file creation takes place. We could just handle this with a `try/catch` block, but I would like to look into it a bit further to avoid possible problems in the future. Unfortunately, I have not been able to reproduce the issue, so could you please include a minimal example? If not possible, some more info on how/when or in which context the empty cache files you listed in the respective [flask-session issue](https://github.com/fengsp/flask-session/issues/132#issuecomment-828724851) are created would be of great help. Also, you mentioned that it happens _sometimes_. How frequent would that be?

**Comment by csm10495** [comment]
Honestly I've only seen it once or twice ever! It's really just saving things in the session object and eventually with requests happening we saw the zero sized files and eventually it died with those errors.

**Comment by northernSage** [comment]
Oh well, let's see what happens with the next release. I think we could leave the issue open for now so people have somewhere to discuss this if they come across the same problem and hopefully some more info will come to light. Possible reasons that come to mind are something else writing empty files to the cache dir. The file names are md5 hashes, though, which matches the naming convention of `cachelib`. Maybe the processes erring out for other reasons after the file has been created and right before writing to it, thus leaving the empty files behind. All just guessing until I'm able to reproduce this, of course. I'll write the PR adding a `try/catch` block soon, possibly handling this by removing the empty file or ignoring it and showing a message to let the user know. If maintainers or anyone else has a better suggestions here, feel free to jump in and let me know 👍🏻

**Comment by northernSage** [comment]
`IOError` has been removed in python 3 and it's now an alias for `OSError`. Details in [PEP 3151](https://www.python.org/dev/peps/pep-3151/#confusing-set-of-os-related-exceptions). Since the latest release already handles `OSError` in [cachelib.FileSystemCache._prune ](https://github.com/pallets/cachelib/blob/main/src/cachelib/file.py#L95) I will be closing this for now. Feel free to reopen if you see the same behaviour again 👍🏻

**Comment by northernSage** [comment]
> The same bug described with cachlib version 0.1.1, seems to appear also in version 0.2.0
> 
> I faced the same issue with 0.1.1:
> 
> [Tue Jul 06 09:57:39.680042 2021] [wsgi:error] [pid 172159] [client 127.0.0.1:38240] self._prune()
> [Tue Jul 06 09:57:39.680048 2021] [wsgi:error] [pid 172159] [client 127.0.0.1:38240] File "/lib/python3.6/site-packages/flask_app/cachelib/file.py", line 96, in _prune
> [Tue Jul 06 09:57:39.680053 2021] [wsgi:error] [pid 172159] [client 127.0.0.1:38240] expires = pickle.load(f)
> [Tue Jul 06 09:57:39.680059 2021] [wsgi:error] [pid 172159] [client 127.0.0.1:38240] EOFError: Ran out of input
> 
> After replacing cachlib with the new version 0.2.0, this error appears again:
> 
> [Thu Jul 08 12:41:55.193613 2021] [wsgi:error] [pid 111282] [client 127.0.0.1:51574] self._prune()
> [Thu Jul 08 12:41:55.193626 2021] [wsgi:error] [pid 111282] [client 127.0.0.1:51574] File "/lib/python3.6/site-packages/flask_app/cachelib/file.py", line 122, in _prune
> [Thu Jul 08 12:41:55.193639 2021] [wsgi:error] [pid 111282] [client 127.0.0.1:51574] self._remove_expired(now)
> [Thu Jul 08 12:41:55.193652 2021] [wsgi:error] [pid 111282] [client 127.0.0.1:51574] File "/lib/python3.6/site-packages/flask_app/cachelib/file.py", line 91, in _remove_expired
> [Thu Jul 08 12:41:55.193666 2021] [wsgi:error] [pid 111282] [client 127.0.0.1:51574] expires = pickle.load(f)
> [Thu Jul 08 12:41:55.193679 2021] [wsgi:error] [pid 111282] [client 127.0.0.1:51574] EOFError: Ran out of input
> 
> In both cases, the error is in picke.load(f)
> 
> Environment:
> 
> * Python version: python3.6

**Comment by northernSage** [comment]
@bill0070 could you please try cleaning the cache directory and see if it helps or if the empty files are created again? Also, if you could give a bit more context it would help. For example, how did it start to happen (any notable event preceding it?) or how frequent it is

**Comment by bill0070** [comment]
Hi @northernSage,

I cleared the cache by removing all the files and the issue was instantly fixed.
It is an issue that appears randomly, although since it appears it persists until the cache is cleared.
The last time I did the cache clear, there were 2 files and none of them was empty as I checked them before the deletion.
After the clear, new files are being created in the respective path.

I have not observed a specific pattern or a specific action that could result in the issue. 

Since yesterday that I did the latest cache clear I have not yet faced this issue.

Should I try any other actions in order to help to investigate further the issue?
Moreover, when the issue appears again, are there any further actions that I should do at this phase to help the investigation?

Thank you,
BR,
Vasilis

**Comment by northernSage** [comment]
It seems the problem isn't just due to empty files as first suspected, but most surely related to file content since clearing the directory fixes the problem. That is useful information, thanks @bill0070  👍🏻 

> Moreover, when the issue appears again, are there any further actions that I should do at this phase to help the investigation?

The ideal would be to find a way to reproduce this so I can debug it from my side, but since this seems like an intermittent issue I won't ask you for a minimal example. One way to help further, though (if you have the time of course), would be to isolate one of these irregular/corrupt files and share it here using something like https://pastebin.com/ (asumming it doesn't contain any private information). A simple way to isolate one of these files would be:

1. Make a copy of all cache files to a separate directory
2. Remove a single cache file from the original cache directory
3. Check if the issue was solved (does it still raise `EOFError`?)
4. Repeat 2 and 3 until the problem is solved
5. Share the copy of the last removed file

That would give me some more info to find any other reasons for `EOFError` and incorporate fixes to #51

**Comment by bill0070** [comment]
Hi @northernSage,

The issue appeared again and as you suggested, I executed the actions mentioned above.

In more detail, when the issue appeared, the cache files status was like below:

$ ls -l
-rw------- 1 root root    17 Jul 14 16:41 2029240f6d1128be89ddc32729463129
-rw------- 1 root root    91 Jul 14 09:42 570e7190bb00d22f16366a793f53a385
-rw------- 1 root root    91 Jul 14 09:42 66ded671bf5e680ca26afb93a4eee0a0
-rw------- 1 root root   252 Jul 14 04:40 a0e7c44bf5acf7ac6d73b7054c8edec0
-rw------- 1 root root    91 Jul 14 09:42 c90ad92a385a30512eae1358cc282c9d

I tried to delete the files and check then if the issue appears. I started with file "2029240f6d1128be89ddc32729463129".
When I tried to see if the error existed after the file deletion, I noticed that the error was fixed and the status of the cache files was this:

$ ls -l
-rw------- 1 root root    17 Jul 14 16:42 2029240f6d1128be89ddc32729463129
-rw------- 1 root root    91 Jul 14 09:42 570e7190bb00d22f16366a793f53a385
-rw------- 1 root root    91 Jul 14 09:42 66ded671bf5e680ca26afb93a4eee0a0
-rw------- 1 root root    91 Jul 14 16:42 9a99f474a72968d864b9552f0c67ac5a
-rw------- 1 root root   252 Jul 14 04:40 a0e7c44bf5acf7ac6d73b7054c8edec0
-rw------- 1 root root    91 Jul 14 09:42 c90ad92a385a30512eae1358cc282c9d

As you can see above, file "2029240f6d1128be89ddc32729463129" was reacreated.
Although, it seems that there is a difference between these files.

I have attached the old and the new file that seems to create the issue:

[cachelib_error.zip](https://github.com/pallets/cachelib/files/6816617/cachelib_error.zip)

Thank you,
BR,
Vasilis

**Comment by northernSage** [comment]
This hash is actually a meta file used to keep track of the number of cache entries inside `FileSystemCache`. It was getting removed due to a bug and recreated with the value "0". I could recognise the md5 from previous work on cachelib:

```python
print(md5("__wz_cache_count".encode("utf-8")).hexdigest())
> '2029240f6d1128be89ddc32729463129'
```
`__wz_cache_count` is the name of the internal file in question

#52 addresses this, hopefully it will be permanently fixed in `0.3.0`

Thanks @bill0070!
